hal/src/sercom/v2/uart/reg.rs

```diff
@@ -0,0 +1,411 @@
+        self.usart()
+            .ctrlb
+            .modify(|_, w| w.sbmode().bit(stop_bits.into()));
+    }
+
+    /// Get the current stop bit setting
+    #[inline]
+    pub(super) fn get_stop_bits(&self) -> StopBits {
+        self.usart().ctrlb.read().sbmode().bit().into()
+    }
+
+    /// Enable or disable the start of frame detector.
+    ///
+    /// When set, the UART will generate interrupts for
+    /// RXC and/or RXS if these interrupt flags have been enabled.
+    #[inline]
+    pub(super) fn set_start_of_frame_detection(&mut self, enabled: bool) {
+        self.usart().ctrlb.modify(|_, w| w.sfde().bit(enabled));
+    }
+
+    /// Get the current SOF detector setting
+    #[inline]
+    pub(super) fn get_start_of_frame_detection(&self) -> bool {
+        self.usart().ctrlb.read().sfde().bit()
+    }
+
+    /// Enable or disable the collision detector.
+    ///
+    /// When set, the UART will detect collisions and update the
+    /// corresponding flag in the STATUS register.
+    #[inline]
+    pub(super) fn set_collision_detection(&mut self, enabled: bool) {
+        self.usart().ctrlb.modify(|_, w| w.colden().bit(enabled));
+    }
+
+    /// Get the current collision detector setting
+    #[inline]
+    pub(super) fn get_collision_detection(&self) -> bool {
+        self.usart().ctrlb.read().colden().bit()
+    }
+
+    /// Set the baud rate
+    ///
+    /// This function will calculate the best BAUD register setting based on the
+    /// stored GCLK frequency and desired baud rate. The maximum baud rate is
+    /// GCLK frequency/oversampling. Values outside this range will saturate at
+    /// the maximum supported baud rate.
+    ///
+    /// Note that 3x oversampling is not supported.
+    #[inline]
+    pub(super) fn set_baud<B: Into<Hertz>>(&mut self, freq: Hertz, baud: B, mode: BaudMode) {
+        let baud: Hertz = baud.into();
+        let usart = self.usart();
+
+        usart
+            .ctrla
+            .modify(|_, w| unsafe { w.sampr().bits(mode.sampr()) });
+
+        match mode {
+            BaudMode::Arithmetic(n) => {
+                let baud = calculate_baud_asynchronous_arithm(baud.0, freq.0, n as u8);
+                unsafe { usart.baud_usartfp_mode().write(|w| w.baud().bits(baud)) };
+            }
+
+            BaudMode::Fractional(n) => {
+                let (baud, frac) = calculate_baud_asynchronous_fractional(baud.0, freq.0, n as u8);
+                unsafe {
+                    usart.baud_frac_mode().write(|w| {
+                        w.fp().bits(frac);
+                        w.baud().bits(baud)
+                    });
+                }
+            }
+        };
+    }
+
+    /// Get the contents of the `BAUD` register and the current baud mode. Note
+    /// that only the CONTENTS of `BAUD` are returned, and not the actual baud
+    /// rate. Refer to the datasheet to convert the `BAUD` register contents
+    /// into a baud rate.
```

Comment:
Just out of curiosity, why not do the conversion for users? Do you think the raw pieces are more useful? Or would the actual frequency be more useful? I'm imagining a situation where you might want to set the desired baud rate, then read it back to see how close the actual baud rate is.

Reply:
The only reason I return the raw contents of the `BAUD` register is to avoid having to hold the underlying GCLK frequency in `Config`. I'm trying to avoid synchronization issues where the clock frequency gets changed from under the UART. Thankfully this should be impossible when `clock::v2` gets here.

Reply:
Yes, with `clock::v2`, the `Config` will hold a `Pclk` (peripheral channel clock) struct. Holding that struct will guarantee that the GCLK doesn't change, and it also contains a copy of the GCLK frequency.

Maybe you can update this once `clock::v2` lands.